p2p/test/transport/integration_test.go

```diff
@@ -0,0 +1,290 @@
+			} else {
+				libp2pOpts = append(libp2pOpts, libp2p.ListenAddrStrings("/ip4/127.0.0.1/udp/0/quic-v1/webtransport"))
+			}
+			h, err := libp2p.New(libp2pOpts...)
+			require.NoError(t, err)
+			return h
+		},
+	},
+}
+
+func TestSmallStream(t *testing.T) {
+	for _, tc := range transportsToTest {
+		t.Run(tc.Name, func(t *testing.T) {
+			h1 := tc.HostGenerator(t, TranspotTestCaseOpts{})
+			h2 := tc.HostGenerator(t, TranspotTestCaseOpts{NoListen: true})
+
+			require.NoError(t, h2.Connect(context.Background(), peer.AddrInfo{
+				ID:    h1.ID(),
+				Addrs: h1.Addrs(),
+			}))
+
+			ctx := context.Background()
+			res := <-ping.Ping(ctx, h2, h1.ID())
+			require.NoError(t, res.Error)
+		})
+	}
+}
+
+func TestBigPing(t *testing.T) {
+	// 64k buffers
+	sendBuf := make([]byte, 64<<10)
+	recvBuf := make([]byte, 64<<10)
+	totalSends := 64
+
+	// Fill with random bytes
+	_, err := rand.Read(sendBuf)
+	require.NoError(t, err)
+
+	for _, tc := range transportsToTest {
+		t.Run(tc.Name, func(t *testing.T) {
+			h1 := tc.HostGenerator(t, TranspotTestCaseOpts{})
+			h2 := tc.HostGenerator(t, TranspotTestCaseOpts{NoListen: true})
+
+			require.NoError(t, h2.Connect(context.Background(), peer.AddrInfo{
+				ID:    h1.ID(),
+				Addrs: h1.Addrs(),
+			}))
+
+			h1.SetStreamHandler("/BIG-ping/1.0.0", func(s network.Stream) {
+				io.Copy(s, s)
+				s.Close()
+			})
+
+			errCh := make(chan error, 1)
+			allocs := testing.AllocsPerRun(10, func() {
+				s, err := h2.NewStream(context.Background(), h1.ID(), "/BIG-ping/1.0.0")
+				require.NoError(t, err)
+				defer s.Close()
+
+				go func() {
+					for i := 0; i < totalSends; i++ {
+						_, err := io.ReadFull(s, recvBuf)
+						if err != nil {
+							errCh <- err
+							return
+						}
+						if !bytes.Equal(sendBuf, recvBuf) {
+							errCh <- fmt.Errorf("received data does not match sent data")
+						}
+
+					}
+					_, err = s.Read([]byte{0})
+					errCh <- err
+				}()
+
+				for i := 0; i < totalSends; i++ {
+					s.Write(sendBuf)
+				}
+				s.CloseWrite()
+				require.ErrorIs(t, <-errCh, io.EOF)
+			})
+
+			if int(allocs) > (len(sendBuf)*totalSends)/4 {
+				t.Logf("Expected fewer allocs, got: %f", allocs)
+			}
+		})
+	}
+}
+
+func TestManyStreams(t *testing.T) {
+	streamCount := 128
+	for _, tc := range transportsToTest {
+		t.Run(tc.Name, func(t *testing.T) {
+			h1 := tc.HostGenerator(t, TranspotTestCaseOpts{NoRcmgr: true})
+			h2 := tc.HostGenerator(t, TranspotTestCaseOpts{NoListen: true, NoRcmgr: true})
+
+			require.NoError(t, h2.Connect(context.Background(), peer.AddrInfo{
+				ID:    h1.ID(),
+				Addrs: h1.Addrs(),
+			}))
+
+			h1.SetStreamHandler("echo", func(s network.Stream) {
+				io.Copy(s, s)
+				s.CloseWrite()
+			})
+
+			streams := make([]network.Stream, streamCount)
+			for i := 0; i < streamCount; i++ {
+				s, err := h2.NewStream(context.Background(), h1.ID(), "echo")
+				require.NoError(t, err)
+				streams[i] = s
+			}
+
+			wg := sync.WaitGroup{}
```

Comment:
Use an errgroup?

Reply:
TIL! I think I'll avoid it for now since the waitgroup is a bit simpler for this. But thanks for the pointer